internal/kallsyms/kallsyms.go

```diff
@@ -221,70 +219,76 @@ func assignAddresses(f io.Reader, symbols map[string]uint64) error {
 		return nil
 	}
 	r := newReader(f)
-	for r.Line() {
-		s, err, skip := parseSymbol(r, nil)
+	for s, err := range parseSymbol(r, nil) {
 		if err != nil {
-			return fmt.Errorf("parsing kallsyms line: %w", err)
-		}
-		if skip {
-			continue
+			return err
 		}
 
-		existing, requested := symbols[s.name]
+		existing, requested := symbols[string(s.name)]
 		if existing != 0 {
 			// Multiple addresses for a symbol have been found. Return a friendly
 			// error to avoid silently attaching to the wrong symbol. libbpf also
 			// rejects referring to ambiguous symbols.
 			return fmt.Errorf("symbol %s(0x%x): duplicate found at address 0x%x: %w", s.name, existing, s.addr, errAmbiguousKsym)
 		}
 		if requested {
-			symbols[s.name] = s.addr
+			symbols[string(s.name)] = s.addr
 		}
 	}
-	if err := r.Err(); err != nil {
-		return fmt.Errorf("reading kallsyms: %w", err)
-	}
 
 	return nil
 }
 
 type ksym struct {
 	addr uint64
-	name string
-	mod  string
+	name []byte
+	mod  []byte
 }
 
 // parseSymbol parses a line from /proc/kallsyms into an address, type, name and
 // module. Skip will be true if the symbol doesn't match any of the given symbol
 // types. See `man 1 nm` for all available types.
 //
-// Example line: `ffffffffc1682010 T nf_nat_init  [nf_nat]`
-func parseSymbol(r *reader, types []rune) (s ksym, err error, skip bool) {
-	for i := 0; r.Word(); i++ {
-		switch i {
-		// Address of the symbol.
-		case 0:
-			s.addr, err = strconv.ParseUint(r.Text(), 16, 64)
-			if err != nil {
-				return s, fmt.Errorf("parsing address: %w", err), false
+// Example line: `ffffffffc1682010 T nf_nat_init\t[nf_nat]`
+func parseSymbol(r *reader, types []rune) iter.Seq2[ksym, error] {
```

Comment:
Since it's now an iterator:
```suggestion
func symbols(r *reader, types []rune) iter.Seq2[ksym, error] {
```